### ISSUE TITLE
Garbage calibration data is written when there is no hierarchy information

### ISSUE THREAD

**Issue by biochem-fan** [taw10/crystfel#9]
When there is no detector hierarchy information in a geometry file, `indexamajig --mille` rightly complains:

> Cannot write calibration data (--mille) because geometry does not contain hierarchy information (group 'all' not found)

However, non zero-byte mille files are actually written. Passing this to `align-detector` causes a segmentation fault.

> If the detector consists of only one panel, CrystFEL will automatically create the all group containing it.
https://gitlab.desy.de/thomas.white/crystfel/-/blob/master/doc/man/crystfel_geometry.5.md

How about assuming all panels belong to the `all` group if there is no definition?

**Comment by biochem-fan** [comment]
Thank you very much for your quick fix!

> Technically it's allowed, but it causes problems as you noticed.

Would it be difficult to make mille file writing more robust (or atomic)?

Another error case is when `indexamajig` crashed or terminated early by Ctrl-C. This also creates corrupted mille file and breaks `align_detector`. Sometimes it is not trivial to find which file is responsible.

> I like the idea of automatically adding an all group with all panels, if nothing else is defined. I made a note of it over here: https://gitlab.desy.de/thomas.white/crystfel/-/issues/96

This will have an additional benefit that we can use the same geometry file for version 0.11 and earlier at least for single-level detectors. If we add "group_all", older versions of CrystFEL cannot load the geometry, hitting [this check](https://github.com/taw10/crystfel/blob/a38407fe140d6336e2fa02f0240b1ec993ca321a/libcrystfel/src/datatemplate.c#L1239-L1244). My pipeline at SACLA has to generate geometry files compatible with various versions of CrystFEL, so this change would be very welcome.

**Comment by taw10** [comment]
> Would it be difficult to make mille file writing more robust (or atomic)?
> 
> Another error case is when `indexamajig` crashed or terminated early by Ctrl-C. This also creates corrupted mille file and breaks `align_detector`. Sometimes it is not trivial to find which file is responsible.

Indeed, the incomplete/corrupted files are already a big problem.  I hope to improve this as part of making `indexamajig` write only one Mille file per job.  See https://gitlab.desy.de/thomas.white/crystfel/-/issues/92.
